### ISSUE TITLE
Missing CORS Rule ID

### ISSUE THREAD

**Issue by prem-prakash** [aws/aws-sdk-ruby#2484]
`Aws::S3::Types::CORSRule` 
https://github.com/aws/aws-sdk-ruby/blob/master/gems/aws-sdk-s3/lib/aws-sdk-s3/types.rb#L689
do not have the ID field

The id field is present in other languages SDK like .NET https://github.com/aws/aws-sdk-net/blob/master/sdk/src/Services/S3/Custom/Model/CORSRule.cs#L72

It is important to identify the rule.

**Comment by mullermp** [comment]
Hey, thanks for opening an issue. These types are generated from the s3 service model: https://github.com/aws/aws-sdk-ruby/blob/34f240beff5219df268fec68065374b47d894c23/apis/s3/2006-03-01/api-2.json#L1259. The s3 api docs do not mention an ID for this structure. I do know that the .NET s3 client is hand written. That ID shouldn't exist at all but is likely there for legacy support. Could you please share your use case and what this ID is doing?

**Comment by prem-prakash** [comment]
Hi @mullermp thanks for answering this issue.

In my use case, we use the ID to define a default CORSRule, which is updated separated from the other ones. We use S3 CORS rules to allow our clients' domains to access S3 content. When an account of a client is created we update the allowed origins of the default (identified by the ID) CORSRule. The other CORS rules are defined for other purposes.

Of course, we can make a detection using Rule content to identify it but is a bit more complex than having an ID to deal with.
In summary, CORS is a list of CORSRule, being able to set an ID is very useful to identify each CORSRule by this ID.

**Comment by mullermp** [comment]
I see. It's not part of S3's official API (in which our code is generated around). I've reached out to the .NET SDK guys to see why that was added in the first place. I don't know if we can augment the `CORSRule` type and if we could I'm not sure whether or not we should even do that. We try to stay true to the API reference. For S3, .NET (and Java V1) are exceptions because they are hand written clients, predating code generated SDKs.

Regarding your use case, I think you could easily wrap the CORSRule parameters in a Ruby hash like so:
```
# can use any ID you like, as a hash key
rules = {
  'id-1': { allowed_headers: [ "Authorization" ], allowed_methods: [ "GET" ] },
  'id-2': { allowed_headers: [ "Authorization" ], allowed_methods: [ "POST" ] },
  ...
}
```

And then when you call your operation:
```
resp = client.put_bucket_cors(
  bucket: bucket, 
  cors_configuration: {
    cors_rules: [ rules['id-1'], rules['id-2'] ]
  }
)
```

**Comment by prem-prakash** [comment]
The ID field was present in the V1 of the ruby SDK:

https://github.com/aws/aws-sdk-ruby/blob/1.6.7/lib/aws/s3/cors_rule.rb#L60

I am facing this issue due to the upgrade to V3. The .NET SDK was just an example.

Regarding your proposed solution, it would work to put new cors rules, but the problem is when I want to search for a specific rule to update it.

I found another way to handle my use case. I am just finding the rule by its attributes, fields allowed_headers and allowed_methods, and then updating it.

So we can close this one.

Anyway thanks for the support.

**Comment by mullermp** [comment]
Ah! Weird that this is in v1 of the SDK.. v1 Ruby SDK was also hand written (v2/v3 are code generated) but I left it out because it's long been deprecated, in 2015. This is really strange. Thank you for pointing this out. I wonder if there is a hidden S3 feature. I'm going to investigate further.

**Comment by mullermp** [comment]
https://docs.aws.amazon.com/AmazonS3/latest/userguide/enabling-cors-examples.html

This example has usage of Rule ID! I think you are on to something. I think I need to engage an appropriate S3 team and find out why this is not modeled.

**Comment by prem-prakash** [comment]
Yeap, I saw these examples too. Hope to hear from this interaction with an S3 team.

**Comment by mullermp** [comment]
I've engaged the appropriate team. This may take a while. The plan is to get this confirmed by the S3 team, find out why it's missing (it could just be a mistake), and if that's the case, get it included into an SDK release for all the languages.

**Comment by mullermp** [comment]
I'm working with S3 and the documentation writers to prepare a release for this, targeting next week.

**Comment by mullermp** [comment]
This has been added: https://github.com/aws/aws-sdk-ruby/blob/0465bfacbf87e6bc78c38191961ed860413d85cd/apis/s3/2006-03-01/api-2.json#L1266
